Make kargs.sh work with systemd service and for Ubuntu

e0ne opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved kernel-argument editing by selecting the correct script context dynamically, ensuring commands run in the right environment.
  * Enhanced Ubuntu GRUB handling by safely modifying GRUB_CMDLINE_LINUX_DEFAULT and only applying changes when needed.
  * Added host script placement plus SELinux file context labeling for reliable kernel-arg updates in systemd mode.
* **Tests**
  * Updated daemon controller test setup to support Bash execution during command runs.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ykulazhenkov commented on pkg/plugins/generic/generic_plugin.go:
```diff
@@ -81,7 +81,8 @@ type genericPluginOptions struct {
 
-const scriptsPath = "bindata/scripts/kargs.sh"
+const daemonScriptsPath = "bindata/scripts/kargs.sh"
+const systemdScriptsPath = "//var/lib/sriov/kargs.sh"
```
`//` at the beginning

ykulazhenkov commented on pkg/plugins/generic/generic_plugin.go:
```diff
@@ -277,7 +278,11 @@ func needDriverCheckVdpaType(state *sriovnetworkv1.SriovNetworkNodeState, driver
 // editKernelArg Tries to add the kernel args via ostree or grubby.
 func editKernelArg(helper helper.HostHelpersInterface, mode, karg string) error {
```
Will this logic really work in the systemd mode? How it will behaves?
1. Reboot
2. systemd service exec
3. systemd service edited the grub config
4. Host booted
5. sriov-daemon pod starts
6. sriov-daemon detects that running kernel is missing required args
7.  host reboot again
8. ...
9. stable state

Is the sequence above describes the flow properly?

e0ne replied:
in systemd mode kernel arg are configured by sriov-service so it doesn't do reboot loop

adrianchiris replied:
@e0ne wont kernel args be applied only after reboot ?
hence two reboots will be required for this case ? 

(unless we set kargs in daemon even in systemd mode)

ykulazhenkov commented on bindata/manifests/daemon/daemonset.yaml:
```diff
@@ -116,7 +116,7 @@ spec:
             - /bin/bash
             - -c
-            - mkdir -p /host/var/lib/sriov/ && cp /usr/bin/sriov-network-config-daemon /host/var/lib/sriov/sriov-network-config-daemon && chcon -t bin_t /host/var/lib/sriov/sriov-network-config-daemon | true # Allow systemd to run the file, use pipe true to not failed if the system doesn't have selinux or apparmor enabled
+            - mkdir -p /host/var/lib/sriov/ && cp /usr/bin/sriov-network-config-daemon /host/var/lib/sriov/sriov-network-config-daemon && chcon -t bin_t /host/var/lib/sriov/sriov-network-config-daemon | true  && cp /bindata/scripts/kargs.sh /host/var/lib/sriov/kargs.sh && chcon -t bin_t /host/var/lib/sriov/kargs.sh | true # Allow systemd to run the file, use pipe true to not failed if the system  ...
```
I think the original command has issue. Should it has `||` (or) instead of `|` (pipe)? The same question for the changes you added.

e0ne replied:
yes, we need to fix it and we should we need to think how to simplify this line. thanks for pointing me on this!

e0ne replied:
just tested again: we need only one `|` to not fail if `chcon` execution failed

ykulazhenkov replied:
Yes, this will work because `|` will mask errors from the previous command (by default, if `set -o pipefail` is not set). The `||` looks cleaner to me, but I'm okay to keep the current implementation.

adrianchiris commented on bindata/scripts/kargs.sh:
```diff
@@ -5,54 +5,83 @@ command=$1
-args=$(chroot /host/ cat /proc/cmdline)
 
-IS_OS_UBUNTU=true; [[ "$(chroot /host/ grep -i ubuntu /etc/os-release -c)" == "0" ]] && IS_OS_UBUNTU=false
+if [[ -d "/bindata/scripts" ]];then
```
nit :add comment, its non obvious that in this case we run in from config daemon and the other by system service.

e0ne replied:
done

zeeke replied:
can you check the comment again.  it might have been dropped by a git rebase

SchSeba replied:
I don't see the comment also

adrianchiris commented on bindata/scripts/kargs.sh:
```diff
@@ -5,54 +5,83 @@ command=$1
+
+args=$(chroot "$chroot_path" cat /proc/cmdline)
+
+IS_OS_UBUNTU=true; [[ "$(chroot "$chroot_path" grep -i ubuntu /etc/os-release -c)" == "0" ]] && IS_OS_UBUNTU=false
```
you can:

```
. /etc/os-release
if [[ $ID == "ubuntu" ]]; then
..
fi
```

e0ne replied:
done

e0ne replied:
rolled back to a previous solution. this doesn't work for chroot

adrianchiris reviewed (commented):
overall LGTM. some minor nits.

e0ne committed:
Make kargs.sh work with systemd service and for Ubuntu

Signed-off-by: Ivan Kolodiazhnyi <ikolodiazhny@nvidia.com>

zeeke commented on bindata/manifests/daemon/daemonset.yaml:
```diff
@@ -125,6 +125,8 @@ spec:
               cp /usr/bin/sriov-network-config-daemon /host/var/lib/sriov/sriov-network-config-daemon
               chcon -t bin_t /host/var/lib/sriov/sriov-network-config-daemon || true # Allow systemd to run the file, use pipe true to not failed if the system doesn't have selinux or apparmor enabled
+              cp /bindata/scripts/kargs.sh /host/var/lib/sriov/kargs.sh
+              chcon -t bin_t /host/var/lib/sriov/kargs.sh | true # Allow systemd to run the file, use pipe true to not failed if the system doesn't have selinux or apparmor enabled
```
nit: mind making congruent to L127?
```suggestion
              chcon -t bin_t /host/var/lib/sriov/kargs.sh || true # Allow systemd to run the file, use pipe true to not failed if the system doesn't have selinux or apparmor enabled
```

alekseysen committed:
Add Ubuntu support

Many customers rely on the Ubuntu operating system, and the ability
to automatically update GRUB parameters is essential.

Signed-off-by: Aleksey Senin <alekseys@nvidia.com>

ykulazhenkov commented on bindata/scripts/kargs.sh:
```diff
@@ -60,4 +119,4 @@ else
 fi
 
-echo $ret
+echo $ret
```
nit: no new line at the end

e0ne replied:
fixed

adrianchiris commented on bindata/scripts/kargs.sh:
```diff
@@ -9,8 +9,67 @@ args=$(chroot /host/ cat /proc/cmdline)
+                fi
+            done
+
+            if [ $found == false ];then
```
what will happen if we have a karg already set with a certain value and we want to change to a different value ?

example:
current grub cmdline args have: `ib_core.netns_mode=1` then we want to set (using kargs.sh) `ib_core.netns_mode=0`

in this case we will end up with both set in grub cmdline IIUC since we compare the full key=val

i think we have the same issue with rpm-ostree as well as we use only append and delete.

imo we need to seach for the key of the karg first, if exists and its value is different then we remove and readd with new value (rpm-ostree has --replace flag)

e0ne replied:
Let's address it in a follow-up PR both for Ubuntu and RHEL-based OSes

SchSeba commented:
Hi @e0ne,

I must say I don't follow this change.
we copy the kargs.sh into the host that is fine so when the service runs it will be able to call it.

but inside the script we still have pointers to `/host` and that will not work.
I think we need to add a check to see if we are inside the container or not.
maybe try to run a `chroot` and if that failed we know we are on the host and we need to changes stuff like `cp /host/${grub_config}`

/hold

e0ne commented:
> Hi guys, what about adding debian support? Looks like debian is configured similarly :)

We don't test Debian but I suppose it should work. We'll be glad to review PR with Debian support

SchSeba commented on pkg/plugins/generic/generic_plugin.go:
```diff
@@ -269,7 +271,11 @@ func needDriverCheckVdpaType(state *sriovnetworkv1.SriovNetworkNodeState, driver
+	if _, err := os.Stat(consts.Host); errors.Is(err, os.ErrNotExist) {
+		script = systemdScriptsPath
+	}
+	_, _, err := helper.RunCommand("/bin/bash", script, mode, karg)
```
can we stay with `sh` here as before? or swtich the other calls to `bash` but please lets be consistent

e0ne replied:
an updated scripts doesn't work with `sh` beacause it uses `bash`. do we want ti switch to `bash` everywhere?

SchSeba commented on pkg/daemon/daemon_test.go:
```diff
@@ -119,72 +112,97 @@ var _ = Describe("Daemon Controller", Ordered, func() {
 		// general
 		hostHelper.EXPECT().Chroot(gomock.Any()).Return(func() error { return nil }, nil).AnyTimes()
 		hostHelper.EXPECT().RunCommand("/bin/sh", gomock.Any(), gomock.Any(), gomock.Any()).Return("", "", nil).AnyTimes()
+		hostHelper.EXPECT().RunCommand("/bin/bash", gomock.Any(), gomock.Any(), gomock.Any()).Return("", "", nil).AnyTimes()
```
please check the comment below about this

SchSeba commented on bindata/scripts/kargs.sh:
```diff
@@ -9,8 +9,67 @@ args=$(chroot /host/ cat /proc/cmdline)
 
 IS_OS_UBUNTU=true; [[ "$(chroot /host/ grep -i ubuntu /etc/os-release -c)" == "0" ]] && IS_OS_UBUNTU=false
```
question how this one will work when we run on systemd mode?
the /host folder doesn't exist in the host itself only inside the container.

e0ne commented on bindata/scripts/kargs.sh:
```diff
@@ -9,8 +9,67 @@ args=$(chroot /host/ cat /proc/cmdline)
 if ${IS_OS_UBUNTU} ; then
+    grub_config="/etc/default/grub"
+    # Operate on the copy of the file
+    cp /host/${grub_config} /tmp/grub
```
fixed

SchSeba commented on bindata/scripts/kargs.sh:
```diff
@@ -1,63 +1,144 @@
 #!/bin/bash
+# Copyright 2025 sriov-network-device-plugin authors
```
I think this copy right is wrong no?

SchSeba commented on bindata/scripts/kargs.sh:
```diff
@@ -1,63 +1,144 @@
+
+    if [ $ret -ne 0 ];then
+        # Update grub only if there were changes
+        cp /tmp/grub ${chroot_path}/${grub_config}
```
same here lets remove the /

SchSeba reviewed (commented):
I left a comment that I think will make the all implementation much easier.

let me know what do you think

e0ne commented on bindata/scripts/kargs.sh:
```diff
@@ -1,58 +1,124 @@
+
+    if [ $ret -ne 0 ];then
+        # Update grub only if there were changes
+        cp /tmp/grub ${chroot_path}/${grub_config}
```
we use chroor for eache command, so this comment isn't correct

SchSeba commented on pkg/plugins/generic/generic_plugin.go:
```diff
@@ -80,7 +81,8 @@ type genericPluginOptions struct {
 }
 
-const scriptsPath = "bindata/scripts/kargs.sh"
+const daemonScriptsPath = "/bindata/scripts/kargs.sh"
```
nit: can you please update this one to be `containerScriptPath` and `hostScriptPath` I think it's cleaner that way

e0ne replied:
done

SchSeba commented on bindata/scripts/kargs.sh:
```diff
@@ -1,58 +1,124 @@
     done
 else
-    chroot /host/ which grubby > /dev/null 2>&1
+    chroot "$chroot_path" which grubby > /dev/null 2>&1
```
question this case will not work for UBUNTU also?
if on ubuntu we have grubby can't we just use the commands to add and remove like we do for coreOS?

e0ne replied:
unfortunately, there's no grubby in official Canonical repos :(

SchSeba commented on bindata/scripts/kargs.sh:
```diff
@@ -1,58 +1,124 @@
 
-IS_OS_UBUNTU=true; [[ "$(chroot /host/ grep -i ubuntu /etc/os-release -c)" == "0" ]] && IS_OS_UBUNTU=false
+if [[ -d "/bindata/scripts" ]];then
+    chroot_path="/host/"
```
in general I think it's better to have `command='chroot /host'` and `command=''` then just use it in all the script

e0ne replied:
IMO, `chroot_path` helps to understand logic better rather then `comand`. also it simplifies commands execution in the script

SchSeba commented on bindata/scripts/kargs.sh:
```diff
@@ -1,58 +1,124 @@
+IS_OS_UBUNTU=true; [[ "$(chroot "$chroot_path" grep -i ubuntu /etc/os-release -c)" == "0" ]] && IS_OS_UBUNTU=false
 
-# Kernel args configuration isn't supported for Ubuntu now, so we shouldn't do anything here
 if ${IS_OS_UBUNTU} ; then
```
please check the comment below and let me know if there is a reason that ubuntu is different from a grubby based system. so we can just use the command to add and remove kernel args

e0ne replied:
there's no cli in oficial Ubuntu repos to modify kernel args

ykulazhenkov commented on bindata/scripts/kargs.sh:
```diff
@@ -1,58 +1,124 @@
 
-# Kernel args configuration isn't supported for Ubuntu now, so we shouldn't do anything here
 if ${IS_OS_UBUNTU} ; then
+    grub_config="/etc/default/grub"
```
Editing /etc/default/grub is unreliable because files in /etc/default/grub.d/*.cfg are loaded afterward and can override its values. Ubuntu cloud images are one common example, as 50-cloudimg-settings.cfg often redefines GRUB_CMDLINE_LINUX_DEFAULT. Could we use a late operator-owned drop-in, such as /etc/default/grub.d/99-sriov-network-operator.cfg, instead?

SchSeba replied:
nice catch I was not aware of that!
+1 on this change

e0ne replied:
done

zeeke commented on pkg/plugins/generic/generic_plugin.go:
```diff
@@ -269,7 +271,11 @@ func needDriverCheckVdpaType(state *sriovnetworkv1.SriovNetworkNodeState, driver
 	log.Log.Info("generic plugin editKernelArg()", "mode", mode, "karg", karg)
-	_, _, err := helper.RunCommand("/bin/sh", scriptsPath, mode, karg)
+	script := containerScriptPath
+	if _, err := os.Stat(consts.Host); errors.Is(err, os.ErrNotExist) {
+		script = hostScriptPath
+	}
```
can you use `vars.UsingSystemdMode` :


```suggestion
    if vars.UsingSystemdMode {
        script = hostScriptPath
    }
```
